test/staking-rewards/MerkleTreeRewards.ts

```diff
@@ -0,0 +1,222 @@
+                        'InvalidClaim'
+                    );
+                });
+
+                it('should revert when calling with an invalid proof', async () => {
+                    const proof = merkleTree.getHexProof(generateLeaf(provider1.address, rewards[provider1.address]));
+
+                    await expect(claim(provider2, { proof })).to.be.revertedWith('InvalidClaim');
+                    await expect(
+                        claim(provider1, {
+                            proof: [formatBytes32String('1234'), ...proof]
+                        })
+                    ).to.be.revertedWith('InvalidClaim');
+                    await expect(claim(provider1, { proof: [...proof].slice(1) })).to.be.revertedWith('InvalidClaim');
+                    await expect(
+                        claim(provider1, {
+                            proof: merkleTree.getHexProof(generateLeaf(provider2.address, rewards[provider1.address]))
+                        })
+                    ).to.be.revertedWith('InvalidClaim');
+                });
+
+                it('should not when claiming twice', async () => {
+                    await claim(provider1);
+                    await expect(claim(provider1)).to.be.revertedWith('AlreadyClaimed');
+                });
+
+                it('should claim the rewards', async () => {
+                    let totalClaimed = BigNumber.from(0);
+
+                    for (const [providerAddress, amount] of Object.entries(rewards)) {
+                        const provider = await ethers.getSigner(providerAddress);
+
+                        expect(await merkleTreeRewards.totalClaimed()).to.equal(totalClaimed);
+                        expect(await merkleTreeRewards.hasClaimed(providerAddress)).to.be.false;
+
+                        const prevTotalSupply = await bnt.totalSupply();
+                        const prevProviderBalance = await bnt.balanceOf(providerAddress);
+                        const prevVaultBalance = await bnt.balanceOf(masterVault.address);
+
+                        const res = await claim(provider);
+
+                        await expect(res)
+                            .to.emit(merkleTreeRewards, stake ? 'RewardsStaked' : 'RewardsClaimed')
+                            .withArgs(providerAddress, amount);
+
+                        totalClaimed = totalClaimed.add(amount);
+
+                        expect(await merkleTreeRewards.totalClaimed()).to.equal(totalClaimed);
+                        expect(await merkleTreeRewards.hasClaimed(providerAddress)).to.be.true;
+
+                        expect(await bnt.totalSupply()).to.equal(prevTotalSupply.add(amount));
+                        expect(await bnt.balanceOf(providerAddress)).to.equal(
+                            prevProviderBalance.add(stake ? 0 : amount)
+                        );
+                        expect(await bnt.balanceOf(masterVault.address)).to.equal(
```

Comment:
the real master vault balance shouldn't change when users stake - maybe mention that it works because it's just a TokenHolder contract.

Reply:
👍